### ISSUE TITLE
Реализовать построение топологии города по данным из QGis.

### ISSUE THREAD

**Issue by GeorgiySurkov** [webebers/city-view-back#25]
В этой задаче сделать только построение графа без лишних вершин, который будет описывать из какого перекрестка можно попасть в какой другой перекресток. Работа с объектами недвижимости производится в другой задаче

**Comment by sevryukov** [comment]
@GeorgiySurkov Это "мега-задача" (фича, а не задача). Нет понимания, какие действия ожидаются от исполнителя, как будет выглядеть техническое решение и как полученный результат будет встроен в конечное решение.
Если вспомнить обсуждаемую концепцию, то конечный результат с точки зрения данных - проект QGIS. Если так, то не совсем из нынешней формулировки понятно, как граф станет частью проекта QGIS, ведь среди поддерживаемых типов пространственных данных нет графа (ибо это и не пространственные данные). Т.е. даже если и говорить о визуализации графа, то это его преобразование в набор геометрических примитивов.

**Comment by GeorgiySurkov** [comment]
@sevryukov, согласен. Заведу две отдельные задачи на построение топологии из данных в qgis с записью в sqlite базу данных и на визуализацию графа из sqlite в qgis с помощью геометрических примитивов.

**Comment by GeorgiySurkov** [comment]
За сегодня пытался настроить подключение к spatialite базе данных из python скрипта. Пробовал QSqlDatabase из PyQT5, но он не видел драйвера, пробовал QgsFeatureRequest.

Завтра попробую sqlite3 из стандартной библиотеки с ручной подгрузкой расширения для работы spatialite.
Затем попробую загрузить эти данные в GeoPandas: https://gist.github.com/perrygeo/868135514d2518257bbb

**Comment by sevryukov** [comment]
@GeorgiySurkov существует несколько зарекомендовавших себя форматов для хранения пространственных данных (и их поддержка для отображения этих данных в QGIS). Почему требуется нечто большее и правильно ли я понимаю, что в качестве этого "нечто большего" выбрана база данных sqlite?

Если я правильно понял новую постановку задачи (кстати, не делайте так - это нарушает смысловую связь между сообщениями в цепочке), то необходимо построить слой дорог и их перекрёстков (эта часть мне понятна), из каких-то данных QGIS. Если так, то вот мои замечания:
1. GQIS - это ПО для обработки пространственных данных, а не источник данных.
2. Проект QGIS может содержать слои с данными и да, функционально возможно создать/построить слой с данными на основе данных других слоёв проекта (материализовав результат в виде сохранённого файла или оставив его виртуальным).
3. OSM может быть источником данных о сети дорог, и эти данные уже представлены как пространственные и готовы к отображению на карте в виде слоя, в частности в QGIS. Я бы предложил обратиться за деталями к Интернет или записям наших семинаров и встреч, проконсультироваться у коллег из второй команды, которая занимается пространственными данными.
4. Преобразовать сеть дорог в граф м.б. ценно для того, чтобы применить к полученным графам алгоритмы для расчёта тех или иных мер, а результат соотнести с уже существующим слоем или представить в виде отдельного слоя.
5. Во всей своей цепочке замечаний и рассуждений я не вижу места для sqlite (как и не вижу из описания задачи и комментариев с Вашей стороны, Георгий).

**Comment by GeorgiySurkov** [comment]
Действительно, я имел в виду в названии задачи OSM вместо QGis. Это опечатка.

**Comment by GeorgiySurkov** [comment]
Что касается sqlite. 

Я решил использовать его после того, как попытался добавить слой с дорожной сетью СПБ. 

Как я пришел к этому решению:
Я получил данные из overpass API в json формате, но qgis с ними не работает из коробки (либо я не разобрался, как это сделать). У меня получилось отобразить эти данные, предварительно переформатировав их в geojson. Получился файл 160 мб. Qgis его мог отображать, но это было слишком медленно. Я нашел информаию в интернете, что лучше преобразовать этот слой в какой-либо поддерживающий индексирование, например, geopackage или spatialite. Выбрав spatialite все заработало очень быстро и плавно. 

Я решил, что было бы удобно хранить топологию города в виде отдельной таблицы в этой же sqlite базе данных, где и хранится таблица с дорожной сетью, полученной из OSM. Поэтому в этой задаче я пробовал сначала подключиться к spatialite бд из python консоли в qgis.

**Comment by GeorgiySurkov** [comment]
Так как из питон консоли работать с spatialite базой у меня до сих пор не получилось, у меня появилась еще одна идея как нам хранить данные. Внутри нашего плагина мы можем работать полностью только с geojson, но не добавлять его в качестве слоя, а упаковывать в geopackage/spatialite. Скорее всего упаковку можно сделать легко с помощью API QGis, так как изначально полученные мной данные в geojson я экспортировал в geopackage и spatialite полностью с помощью qgis. 

Таким образом мы решаем две проблемы: 
- из python работаем с простым форматом (geojson)
- сгенерированные нами слои быстро рендерятся у пользователя, так как сохранены в формате, поддерживающем пространственный индекс.

**Comment by GeorgiySurkov** [comment]
Завел на это задачу #33
